typescript/packages/jumble/src/contexts/AuthenticationContext.tsx

```diff
@@ -0,0 +1,110 @@
+import React, { createContext, useCallback, useContext, useEffect, useState } from "react";
+import { Identity, KeyStore, PassKey } from "@commontools/identity";
+
+// Location in storage of root key.
+const ROOT_KEY = "$ROOT_KEY";
+// "Name" of default persona derived from root key.
+const DEFAULT_PERSONA = "default";
+
+interface AuthenticationContextType {
+  // The authenticated user/persona.
+  user: Identity | void;
+  // Set the authenticated user/persona.
+  authenticate: () => Promise<void>;
+  // Call PassKey registration.
+  registerUser: (name: string, displayName: string) => Promise<void>;
+  // Clears authentication database.
+  clearAuthentication: () => Promise<void>;
+  // Internal: Root key.
+  root: Identity | void;
+  // Internal: Persistent storage for keys. 
+  keyStore: KeyStore | void;
+}
+
+const AuthenticationContext = createContext<AuthenticationContextType>(null!);
+
+// (async) open a key store
+// (async) check to see key exists
+export const AuthenticationProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
+  const [user, setUser] = useState<Identity | void>(undefined);
+  const [keyStore, setKeyStore] = useState<KeyStore | void>(undefined);
+  const [root, setRoot] = useState<Identity | void>(undefined);
+
+  // On load, open the KeyStore and find a root key.
+  useEffect(() => {
+    let ignore = false;
+    async function getKeyStoreAndRoot() {
+      let keyStore = await KeyStore.open();
+      let root = await keyStore.get(ROOT_KEY);
+      if (!ignore) {
+        setKeyStore(keyStore);
+        if (root) {
+          setRoot(root);
+        }
+      }
+    }
+    getKeyStoreAndRoot(); 
+    return () => { ignore = true; }
+  }, []);
+
+  // When root changes, update `user` to the default persona
+  useEffect(() => {
+    let ignore = false;
+    async function setPersona() {
+      setUser(undefined);
+      if (!root) {
+        return;
+      }
+      let user = await root.derive(DEFAULT_PERSONA);
+      if (!ignore) {
+        setUser(user);
+      }
+    }
+    setPersona();
+    return () => { ignore = true; }
+  }, [root]);
+
+  // This should be called when a passkey (possibly) exists for the user already,
+  // and no root key has yet been stored (e.g. first login). Subsequent page loads
+  // should load key from storage and not require this callback.
+  const authenticate = useCallback(async () => {
+    if (!keyStore) {
+      return;
+    }
+    let passkey = await PassKey.get();
+    let root = await passkey.createRootKey();
+    await keyStore.set(ROOT_KEY, root);
+    setRoot(root);
+  }, [keyStore]);
+
+  // This calls out to WebAuthn to register a user. The state of whether
+  // a user has previously registered a passkey is not tracked (and could
+  // only be tracked per document), so this doesn't use any hooks.
+  const registerUser = async (name: string, displayName: string) => {
+    return PassKey.create(name, displayName);
+  };
```

Comment:
Can this one also be `useCallback`'d? Reference stability is important for providers since they can re-render large portions of the app tree accidentally.

Reply:
Would this just be a `useCallback` with `[]` dependencies then?